src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignerInfoTests.netcoreapp.cs

```diff
@@ -0,0 +1,270 @@
+            for (int i = 0; i < array.Count; i++)
+            {
+                if (AsnEncodingDataEqual(array[i], data))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        private static List<AsnEncodedData> GetAllAsnEncodedDataFromAttributes(CryptographicAttributeObjectCollection attributes)
+        {
+            var ret = new List<AsnEncodedData>();
+            foreach (CryptographicAttributeObject attribute in attributes)
+            {
+                foreach (AsnEncodedData attributeValue in attribute.Values)
+                {
+                    ret.Add(attributeValue);
+                }
+            }
+
+            return ret;
+        }
+
+        private static bool AsnEncodingDataEqual(AsnEncodedData a, AsnEncodedData b)
+        {
+            return a.Oid.Value == b.Oid.Value && a.RawData.SequenceEqual(b.RawData);
+        }
+
+        private static SignedCms ReReadSignedCms(SignedCms cms)
+        {
+            byte[] bytes = cms.Encode();
+
+            cms = new SignedCms();
+            cms.Decode(bytes);
+            return cms;
+        }
+
+        private static AsnEncodedData CreateTimestampToken(byte serial)
+        {
+            Oid tokenOid = new Oid(TokenAttributeOid, TokenAttributeOid);
+
+            Oid policyId = new Oid("0.0", "0.0");
+            Oid hashAlgorithmId = new Oid(Oids.Sha256);
+
+            var tokenInfo = new Rfc3161TimestampTokenInfo(
+                policyId,
+                hashAlgorithmId,
+                new byte[256 / 8],
+                new byte[] { (byte)serial },
+                DateTimeOffset.UtcNow);
+
+            return new AsnEncodedData(tokenOid, tokenInfo.Encode());
+        }
+
+        private static byte Read8BitSerial(ReadOnlySpan<byte> serial)
+        {
+            // serial is big endian integer but we expect value in range 0-255
+            byte x = 0;
+            foreach (byte b in serial)
+            {
+                Assert.Equal(0, x);
+                x = b;
+            }
+
+            return x;
+        }
+
+        private static void CheckSignerHasUnsignedAttribute(AsnEncodedData actual, AsnEncodedData expected)
+        {
+            Assert.NotSame(expected.Oid, actual.Oid);
+            Assert.Equal(expected.Oid.Value, actual.Oid.Value);
+            Assert.Equal(expected.Oid.FriendlyName, actual.Oid.FriendlyName);
+
+            // We need to decode bytes because DER and BER may encode the same information slightly differently
+            Rfc3161TimestampTokenInfo expectedToken;
+            Assert.True(Rfc3161TimestampTokenInfo.TryDecode(expected.RawData, out expectedToken, out _));
+
+            Rfc3161TimestampTokenInfo actualToken;
+            Assert.True(Rfc3161TimestampTokenInfo.TryDecode(actual.RawData, out actualToken, out _));
+
+            // serial number can be equivalent but not identical (i.e. serial = 4 can be encoded as [0, 4] or [4])
```

Comment:
The serial number is an ASN Integer value, so it's not allowed to be { 0x00, 0x04 } (The first 9 bits of a value are not allowed to be the same).  Are you seeing otherwise?

(The raw encoded value can be `02 01 04` (BER/CER/DER), `02 81 01 04` (BER), `02 82 00 01 04` (BER), ...; but the payload will always be `04` for the logical concept of "4")

Reply:
It shouldn't be possible to get back a leading zero byte with `04`, but should give one back if the value was, e.g., 0x84.  The AsnReader / AsnSerializer should have thrown in cases where it said "00 04"

"84" and "00 84" are different numbers.  The first is -124, the second is 132.

Reply:
Changing to compare bytes, after talking offline I think differences I've seen was reordering caused by DER/BER and comparing serial number with first byte of the sequence which I assumed was unsigned big-endian large integer - since it was signed I was seeing first byte as 0 and incorrectly put 4 as an example in the comment. Not seeing any issues right now.